src/socket/unix_socket.cr

```diff
@@ -1,50 +1,60 @@
-# A local interprocess communication clientsocket.
+require "./delegates"
+
+# A local interprocess communication (UNIX socket) client socket.
 #
 # Only available on UNIX and UNIX-like operating systems.
 #
-# Example usage:
+# Usage example:
 # ```
 # require "socket"
 #
-# sock = UNIXSocket.new("/tmp/myapp.sock")
-# sock.puts "message"
-# response = sock.gets
-# sock.close
+# UNIXSocket.open("/tmp/myapp.sock") do |socket|
+#   socket.puts "message"
+#   response = socket.gets
+# end
 # ```
-class UNIXSocket < Socket
-  getter path : String?
+class UNIXSocket < IO
+  # Returns the raw socket wrapped by this UNIX socket.
+  getter raw : Socket::Raw
 
-  # Connects a named UNIX socket, bound to a filesystem pathname.
-  def initialize(@path : String, type : Type = Type::STREAM)
-    super(Family::UNIX, type, Protocol::IP)
+  # Creates a `UNIXServer` from a raw socket.
+  def initialize(@raw : Socket::Raw, @address : Socket::UNIXAddress)
+  end
 
-    connect(UNIXAddress.new(path)) do |error|
-      close
+  # Connects a named UNIX socket, bound to a filesystem pathname.
+  def self.new(address : Socket::UNIXAddress) : UNIXSocket
+    base = Socket::Raw.new(Socket::Family::UNIX, Socket::Type::STREAM, Socket::Protocol::IP)
+    base.connect(address) do |error|
+      base.close
       raise error
     end
+    new base, address
   end
 
-  protected def initialize(family : Family, type : Type)
-    super family, type, Protocol::IP
-  end
-
-  protected def initialize(fd : Int32, type : Type, @path : String? = nil)
-    super fd, Family::UNIX, type, Protocol::IP
+  # Connects a named UNIX socket, bound to a filesystem pathname.
+  def self.new(path : String) : UNIXSocket
+    new(Socket::UNIXAddress.new(path))
   end
 
-  # Opens an UNIX socket to a filesystem pathname, yields it to the block, then
-  # eventually closes the socket when the block returns.
+  # Connects a named UNIX socket, bound to a filesystem pathname and yields it to the block.
   #
-  # Returns the value of the block.
-  def self.open(path, type : Type = Type::STREAM)
-    sock = new(path, type)
+  # The socket is closed after the block returns.
+  #
+  # Returns the return value of the block.
+  def self.open(path : Socket::UNIXAddress | String, &block : UNIXSocket ->)
+    socket = new(path)
+
     begin
-      yield sock
+      yield socket
     ensure
-      sock.close
+      socket.close
     end
   end
 
+  Socket.delegate_close
+  Socket.delegate_io_methods
+  Socket.delegate_buffer_sizes
+
   # Returns a pair of unamed UNIX sockets.
```

Comment:
Typo: `unamed` should be `unnamed` and in the [other `self.pair`](https://github.com/crystal-lang/crystal/pull/6929/files#diff-e4de07adb127b6b3c4954da7e7e6b277R92) too.